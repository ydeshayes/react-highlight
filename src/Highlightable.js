import emojiRegex from 'emoji-regex';
import { Component } from 'react';
import PropTypes from 'prop-types';
import EmojiNode from './nodes/EmojiNode';
import Node from './nodes/Node';
import Range from './Range';
import UrlNode from './nodes/UrlNode';
import { getUrl, debounce } from './helpers';

export default class Highlightable extends Component {
  constructor(props) {
    super(props);

    this.dismissMouseUp = 0;
  }

  shouldComponentUpdate(newProps) {
    return newProps.ranges.length !== this.props.ranges.length
      || newProps.text !== this.props.text
      || newProps.enabled !== this.props.enabled;
  }

  getRange(charIndex) {
    return this.props.ranges
      && this.props.ranges.find(range => charIndex >= range.start && charIndex <= range.end);
  }

  onMouseOverHighlightedWord(range, visible) {
    if (visible && this.props.onMouseOverHighlightedWord) {
      this.props.onMouseOverHighlightedWord(range);
    }
  }

  getLetterNode(charIndex, range) {
    return (<Node id={this.props.id}
      range={range}
      charIndex={charIndex}
      key={`${this.props.id}-${charIndex}`}
      highlightStyle={this.props.highlightStyle}>
      {this.props.text[charIndex]}
    </Node>);
  }

  getEmojiNode(charIndex, range) {
    return (<EmojiNode text={this.props.text}
      id={this.props.id}
      range={range}
      key={`${this.props.id}-emoji-${charIndex}`}
      charIndex={charIndex}
      highlightStyle={this.props.highlightStyle} />);
  }

  getUrlNode(charIndex, range, url) {
    return (<UrlNode url={url}
      id={this.props.id}
      range={range}
      key={`${this.props.id}-url-${charIndex}`}
      charIndex={charIndex}
      highlightStyle={this.props.highlightStyle} />);
  }

  mouseEvent() {
    if (!this.props.enabled) {
      return false;
    }

    let text = '';

    if (window.getSelection) {
      text = window.getSelection().toString();
    } else if (document.selection && document.selection.type !== 'Control') {
      text = document.selection.createRange().text;
    }

    if (!text || !text.length) {
      return false;
    }

    const range = window.getSelection().getRangeAt(0);

    const startContainerPosition = parseInt(range.startContainer.parentNode.dataset.position);
    const endContainerPosition = parseInt(range.endContainer.parentNode.dataset.position);

    const startHL = startContainerPosition < endContainerPosition ? startContainerPosition : endContainerPosition;
    const endHL = startContainerPosition < endContainerPosition ? endContainerPosition : startContainerPosition;

    const rangeObj = new Range(startHL, endHL, text, Object.assign({}, this.props, { ranges: undefined }));

    this.props.onTextHighlighted(rangeObj);
  }

  onMouseUp(event) {
    debounce(() => {
      if (this.doucleckicked) {
        this.doucleckicked = false;
        this.dismissMouseUp++;
      } else if (this.dismissMouseUp > 0) {
        this.dismissMouseUp--;
      } else {
        this.mouseEvent.bind(this)();
      }
    }, 200).bind(this)();
  }

  onDoubleClick(event) {
    event.stopPropagation();

    this.doucleckicked = true;
    this.mouseEvent.bind(this)();
  }

  rangeRenderer(letterGroup, range, textCharIndex, onMouseOverHighlightedWord) {
    return this.props.rangeRenderer
      ? this.props.rangeRenderer(letterGroup, range, textCharIndex, onMouseOverHighlightedWord)
      : letterGroup;
  }

  getNode(i, range, text, url, isEmoji) {
    if (url.length) {
      return this.getUrlNode(i, range, url);
    } else if (isEmoji) {
      return this.getEmojiNode(i, range);
    }

    return this.getLetterNode(i, range);
  }

  getRanges() {
    const newText = [];
    let lastRange;

    // For all the characters on the text
    for (let textCharIndex = 0; textCharIndex < this.props.text.length; textCharIndex++) {
      const range = this.getRange(textCharIndex);
      const url = getUrl(textCharIndex, this.props.text);
      const isEmoji = emojiRegex().test(this.props.text[textCharIndex] + this.props.text[textCharIndex + 1]);
      // Get the current character node
      const node = this.getNode(textCharIndex, range, this.props.text, url, isEmoji);

      // If the next node is an url one, we fast forward to the end of it
      if (url.length) {
        textCharIndex += url.length - 1;
      } else if (isEmoji) {
        // Because an emoji is composed of 2 chars
        textCharIndex++;
      }

      if (!range) {
        newText.push(node);
        continue;
      }

      // If the char is in range
      lastRange = range;
      // We put the first range node on the array
      const letterGroup = [node];

      // For all the characters in the highlighted range
      let rangeCharIndex = textCharIndex + 1;

      for (; rangeCharIndex < parseInt(range.end) + 1; rangeCharIndex++) {
        const isEmoji = emojiRegex().test(`${this.props.text[rangeCharIndex]}${this.props.text[rangeCharIndex + 1]}`);

        if (isEmoji) {
          letterGroup.push(this.getEmojiNode(rangeCharIndex, range));
          // Because an emoji is composed of 2 chars
          rangeCharIndex++;
        } else {
          letterGroup.push(this.getLetterNode(rangeCharIndex, range));
        }

        textCharIndex = rangeCharIndex;
      }

      newText.push(this.rangeRenderer(letterGroup,
        range,
        textCharIndex,
        this.onMouseOverHighlightedWord.bind(this)));
    }

    if (lastRange) {
      // Callback function
      this.onMouseOverHighlightedWord(lastRange, true);
    }

    return newText;
  }

  render() {
    const newText = this.getRanges();

    return (
      <div style={this.props.style}
        
        onMouseUp={this.onMouseUp.bind(this)}
        onDoubleClick={this.onDoubleClick.bind(this)}>
        {newText}
      </div>
    );
  }
}

Highlightable.propTypes = {
  ranges: PropTypes.array,
  id: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  text: PropTypes.string,
  enabled: PropTypes.bool,
  onMouseOverHighlightedWord: PropTypes.func,
  onTextHighlighted: PropTypes.func,
  highlightStyle: PropTypes.object,
  style: PropTypes.object,
  rangeRenderer: PropTypes.func
};
