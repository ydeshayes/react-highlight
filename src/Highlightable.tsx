import emojiRegex from 'emoji-regex';
import React from 'react';
import PropTypes from 'prop-types';

import EmojiNode from './nodes/EmojiNode';
import Node from './nodes/Node';
import Range from './Range';
import UrlNode from './nodes/UrlNode';
import {getUrl, debounce} from './helpers';

const Highlightable = (
  {
    ranges,
    onMouseOverHighlightedWord,
    id,
    highlightStyle,
    text,
    enabled,
    rangeRenderer,
    nodeRenderer,
    style,
    onTextHighlighted
  }) => {
  let dismissMouseUp = 0;

  let doucleckicked = false;

  const getRange = charIndex => {
    return ranges
            && ranges.find(r => charIndex >= r.start && charIndex <= r.end);
  };

  const onMouseOverHighlightedWordHandler = (range, visible) => {
    if(visible && onMouseOverHighlightedWord) {
      onMouseOverHighlightedWord(range);
    }
  };

  const getLetterNode = (charIndex, range) => {
    return (<Node id={id}
      range={range}
      charIndex={charIndex}
      key={`${id}-${charIndex}`}
      highlightStyle={highlightStyle}>
      {text[charIndex]}
    </Node>);
  };

  const getEmojiNode = (charIndex, range) => {
    return (<EmojiNode text={text}
      id={id}
      range={range}
      key={`${id}-emoji-${charIndex}`}
      charIndex={charIndex}
      highlightStyle={highlightStyle} />);
  };

  const getUrlNode = (charIndex, range, url) => {
    return (<UrlNode url={url}
      id={id}
      range={range}
      key={`${id}-url-${charIndex}`}
      charIndex={charIndex}
      highlightStyle={highlightStyle} />);
  };

  const mouseEvent = () => {
    if(!enabled) {
      return false;
    }

    let t = '';

    if (window.getSelection) {
      t = window.getSelection().toString();
    } else if (document.selection && document.selection.type !== 'Control') {
      t = document.selection.createRange().text;
    }

    if(!t || !t.length) {
      return false;
    }

    const r = window.getSelection().getRangeAt(0);

    const startContainerPosition = parseInt(r.startContainer.parentNode.dataset.position);
    const endContainerPosition = parseInt(r.endContainer.parentNode.dataset.position);

    const startHL = startContainerPosition < endContainerPosition ? startContainerPosition : endContainerPosition;
    const endHL = startContainerPosition < endContainerPosition ? endContainerPosition : startContainerPosition;

    const rangeObj = new Range(startHL, endHL, text, {
      ranges: undefined,
      onMouseOverHighlightedWord,
      id,
      text,
      enabled,
      rangeRenderer,
      onTextHighlighted,
      style
    });

    onTextHighlighted(rangeObj);
  };

  const onMouseUp = () => {
    debounce(() => {
      if (doucleckicked) {
        doucleckicked = false;
        dismissMouseUp++;
      } else if(dismissMouseUp > 0) {
        dismissMouseUp--;
      } else {
        mouseEvent();
      }
    }, 200)();
  };

  const onDoubleClick = e => {
    e.stopPropagation();

    doucleckicked = true;
    mouseEvent();
  };

  const rangeRendererDefault = (letterGroup, range, textCharIndex, onMouseOverHighlightedWord) => {
    return rangeRenderer
      ? rangeRenderer(letterGroup, range, textCharIndex, onMouseOverHighlightedWord)
      : letterGroup;
  };

  const getNode = (i, r, t, url, isEmoji) => {
    if(nodeRenderer) {
      return nodeRenderer(i, r, t, url, isEmoji);
    }
    
    if(url.length) {
      return getUrlNode(i, r, url);
    } else if(isEmoji) {
      return getEmojiNode(i, r);
    }

    return getLetterNode(i, r);
  };

  const getRanges = () => {
    const newText = [];

    let lastRange;

    // For all the characters on the text
    for(let textCharIndex = 0;textCharIndex < text.length;textCharIndex++) {
      const range = getRange(textCharIndex);
      const url = getUrl(textCharIndex, text);
      const isEmoji = emojiRegex().test(text[textCharIndex] + text[textCharIndex + 1]);
      // Get the current character node
      const node = getNode(textCharIndex, range, text, url, isEmoji);

      // If the next node is an url one, we fast forward to the end of it
      if(url.length) {
        textCharIndex += url.length - 1;
      } else if(isEmoji) {
        // Because an emoji is composed of 2 chars
        textCharIndex++;
      }

      if(!range) {
        newText.push(node);
        continue;
      }

      // If the char is in range
      lastRange = range;
      // We put the first range node on the array
      const letterGroup = [node];

      // For all the characters in the highlighted range
      let rangeCharIndex = textCharIndex + 1;

      for(;rangeCharIndex < parseInt(range.end) + 1;rangeCharIndex++) {
        const isEmoji = emojiRegex().test(`${text[rangeCharIndex]}${text[rangeCharIndex + 1]}`);

        
        if(isEmoji) {
          letterGroup.push(getEmojiNode(rangeCharIndex, range));
          // Because an emoji is composed of 2 chars
          rangeCharIndex++;
        } else {
          letterGroup.push(getNode(rangeCharIndex, range, text, url, isEmoji));
        }

        textCharIndex = rangeCharIndex;
      }

      newText.push(rangeRendererDefault(letterGroup,
        range,
        textCharIndex,
        onMouseOverHighlightedWordHandler));
    }

    if(lastRange) {
      // Callback function
      onMouseOverHighlightedWordHandler(lastRange, true);
    }

    return newText;
  };

  return (
    <div style={style}
      onMouseUp={onMouseUp}
      onDoubleClick={onDoubleClick}>
      {getRanges()}
    </div>
  );
};

Highlightable.propTypes = {
  ranges: PropTypes.array,
  id: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  text: PropTypes.string,
  enabled: PropTypes.bool,
  onMouseOverHighlightedWord: PropTypes.func,
  onTextHighlighted: PropTypes.func,
  highlightStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.func
  ]),
  style: PropTypes.object,
  rangeRenderer: PropTypes.func,
  nodeRenderer: PropTypes.func
};

export default Highlightable;
