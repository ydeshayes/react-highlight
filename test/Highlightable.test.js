import React from 'react';
import sinon from 'sinon';
import { expect as expectChai } from 'chai';
import renderer from 'react-test-renderer';

import Highlightable, { Range, Node } from '../src';

describe('Highlightable component', function () {
  describe('with basic props', function () {
    it('should render the text without highlight', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [];
      const text = 'test the text';

      let component;
      renderer.act(() => {
        component = renderer.create(
          <Highlightable
                Range={range}
                enabled={true}
                onTextHighlighted={onTextHighlighted}
                name={'test'}
                onMouseOverHighlightedWord={onMouseOverHighlightedWord}
                rangeRenderer={(a, b) => b}
                highlightStyle={{
                  backgroundColor: '#ffcc80',
                  enabled: true
                }}
                text={text}
              />,
        )
      });
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 0);
      expectChai(onTextHighlighted).to.have.property('callCount', 0);
    });
  });
  
  describe('with range props', function () {
    it('should render with highlighted text', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [new Range(0, 5)];
      const text = 'test the text';

      let component;
      renderer.act(() => {
        component = renderer.create(
        <Highlightable
            ranges={range}
            enabled={true}
            onTextHighlighted={onTextHighlighted}
            name={'test'}
            onMouseOverHighlightedWord={onMouseOverHighlightedWord}
            rangeRenderer={a => a}
            highlightStyle={{
              backgroundColor: '#ffcc80',
              enabled: true
            }}
            text={text}
          />)
      });

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 1);
      expectChai(onTextHighlighted).to.have.property('callCount', 0);
    });
  });

  describe('testing update', function () {
    it('should highlight text', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [];
      const text = 'test the text';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={a => a}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           text={text}
        />);

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 0);
      expectChai(onTextHighlighted).to.have.property('callCount', 0);

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      const newRange = [new Range(0, 5)];
      renderer.act(() => {
        component.update(<Highlightable
          ranges={newRange}
          enabled={true}
          onTextHighlighted={onTextHighlighted}
          name={'test'}
          onMouseOverHighlightedWord={onMouseOverHighlightedWord}
          rangeRenderer={a => a}
          highlightStyle={{
            backgroundColor: '#ffcc80',
            enabled: true
          }}
          text={text}
       />);
      });

      tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 1);
    });
  });

  describe('with smiley', function () {
    it('should highlight text and keep the smiley at the end of the text', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [];
      const text = 'test the text ðŸ˜˜';

      const component = renderer.create(<Highlightable
        ranges={range}
        enabled={true}
        onTextHighlighted={onTextHighlighted}
        name={'test'}
        onMouseOverHighlightedWord={onMouseOverHighlightedWord}
        rangeRenderer={a => a}
        highlightStyle={{
          backgroundColor: '#ffcc80',
          enabled: true
        }}
        text={text}
     />);

     let tree = component.toJSON();
     expect(tree).toMatchSnapshot();
    });

    it('should highlight text and keep the smiley at the end of the highlighted text', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [new Range(13, 14)];
      const text = 'test the text ðŸ˜˜';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={a => a}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           text={text}
        />);

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 1);
    });

    it('should highlight text and keep the smiley in the middle of the highlighted text', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [new Range(13, 18)];
      const text = 'test the text ðŸ˜˜ test again';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={a => a}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           text={text}
        />);

      
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 1);
    });
  });

  describe('with url', function () {
    it('should render with url', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [];
      const text = 'test http://www.google.fr';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={a => a}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           text={text}
        />);

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render with highlighted url', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const range = [new Range(5, 7)];
      const text = 'test http://www.google.fr';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={a => a}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           text={text}
        />);


      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      
      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 1);
      expectChai(onTextHighlighted).to.have.property('callCount', 0);
    });
  });

  describe('with custom node', function() {
    it('should render the text wit custom node', () => {
      const onMouseOverHighlightedWord = sinon.spy();
      const onTextHighlighted = sinon.spy();
      const nodeRenderer = sinon.spy((i, r, text) => {
        return <Node id="test"
        range={r}
        charIndex={i}
        key={`test-${i}`}>{text[i]}</Node>
      });
      const range = [];
      const text = 'test the text';

      const component = renderer.create(<Highlightable
           ranges={range}
           enabled={true}
           onTextHighlighted={onTextHighlighted}
           name={'test'}
           onMouseOverHighlightedWord={onMouseOverHighlightedWord}
           rangeRenderer={(a, b) => b}
           highlightStyle={{
             backgroundColor: '#ffcc80',
             enabled: true
           }}
           nodeRenderer={nodeRenderer}
           text={text}
        />);

      expectChai(onMouseOverHighlightedWord).to.have.property('callCount', 0);
      expectChai(onTextHighlighted).to.have.property('callCount', 0);
      expectChai(nodeRenderer).to.have.property('callCount', 13);

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});