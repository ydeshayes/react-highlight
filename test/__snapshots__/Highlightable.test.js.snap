// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlightable component testing update should highlight text 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
</div>
`;

exports[`Highlightable component testing update should highlight text 2`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={1}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    e
  </span>
  <span
    data-position={2}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    s
  </span>
  <span
    data-position={3}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={4}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
     
  </span>
  <span
    data-position={5}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
</div>
`;

exports[`Highlightable component with basic props should render the text without highlight 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
</div>
`;

exports[`Highlightable component with custom node should render the text wit custom node 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
</div>
`;

exports[`Highlightable component with range props should render with highlighted text 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={1}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    e
  </span>
  <span
    data-position={2}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    s
  </span>
  <span
    data-position={3}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={4}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
     
  </span>
  <span
    data-position={5}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
</div>
`;

exports[`Highlightable component with smiley should highlight text and keep the smiley at the end of the highlighted text 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
  <span
    data-position={13}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
     
  </span>
  <span
    data-position={14}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    ðŸ˜˜
  </span>
</div>
`;

exports[`Highlightable component with smiley should highlight text and keep the smiley at the end of the text 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
  <span
    data-position={13}
  >
     
  </span>
  <span
    data-position={14}
  >
    ðŸ˜˜
  </span>
</div>
`;

exports[`Highlightable component with smiley should highlight text and keep the smiley in the middle of the highlighted text 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
  >
    t
  </span>
  <span
    data-position={6}
  >
    h
  </span>
  <span
    data-position={7}
  >
    e
  </span>
  <span
    data-position={8}
  >
     
  </span>
  <span
    data-position={9}
  >
    t
  </span>
  <span
    data-position={10}
  >
    e
  </span>
  <span
    data-position={11}
  >
    x
  </span>
  <span
    data-position={12}
  >
    t
  </span>
  <span
    data-position={13}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
     
  </span>
  <span
    data-position={14}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    ðŸ˜˜
  </span>
  <span
    data-position={16}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
     
  </span>
  <span
    data-position={17}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    t
  </span>
  <span
    data-position={18}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
      }
    }
  >
    e
  </span>
  <span
    data-position={19}
  >
    s
  </span>
  <span
    data-position={20}
  >
    t
  </span>
  <span
    data-position={21}
  >
     
  </span>
  <span
    data-position={22}
  >
    a
  </span>
  <span
    data-position={23}
  >
    g
  </span>
  <span
    data-position={24}
  >
    a
  </span>
  <span
    data-position={25}
  >
    i
  </span>
  <span
    data-position={26}
  >
    n
  </span>
</div>
`;

exports[`Highlightable component with url should render with highlighted url 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
    style={
      {
        "backgroundColor": "#ffcc80",
        "enabled": true,
        "wordWrap": "break-word",
      }
    }
  >
    <a
      data-position={25}
      href="http://www.google.fr"
      target="blank"
    >
      http://www.google.fr
    </a>
  </span>
</div>
`;

exports[`Highlightable component with url should render with url 1`] = `
<div
  onDoubleClick={[Function]}
  onMouseUp={[Function]}
>
  <span
    data-position={0}
  >
    t
  </span>
  <span
    data-position={1}
  >
    e
  </span>
  <span
    data-position={2}
  >
    s
  </span>
  <span
    data-position={3}
  >
    t
  </span>
  <span
    data-position={4}
  >
     
  </span>
  <span
    data-position={5}
    style={
      {
        "wordWrap": "break-word",
      }
    }
  >
    <a
      data-position={25}
      href="http://www.google.fr"
      target="blank"
    >
      http://www.google.fr
    </a>
  </span>
</div>
`;
